setupext/build_makefile.py

```diff
@@ -0,0 +1,199 @@
+        i0 = None
+        i1 = None
+        for i,v in enumerate(x):
+            if v == '-c': i1=i
+            elif v == '-o': i0=i
+        pre = set(x[:i1])
+        post = x[i0+2:]
+
+        self.compile_command = command
+        self.compile_pre = pre
+        self.compile_post = post
+        self.includes = includes
+        self.sources.append(x[i1+1])
+
+    def captureLink(self, x):
+        self.link_command = x[0]
+        x = x[1:]
+        self.library = x[-1]
+        print(x[-3:])
+        x = x[:-3]
+        self.objects = [ i for i in x if i.endswith(".o")]
+        self.link_options = [ i for i in x if not i.endswith(".o")]
+        u = self.objects[0].split("/")
+        self.build_dir = "/".join(u[:2])
+
+    def compile(self, *args, **kwargs):
+        self.actual.spawn = self.captureCompile
+        return self.actual.compile(*args, **kwargs)
+
+    def link_shared_object(self, *args, **kwargs):
+        self.actual.spawn = self.captureLink
+        return self.actual.link_shared_object(*args, **kwargs)
+
+    def detect_language(self, x):
+        return self.actual.detect_language(x)
+
+    def write(self):
+        library = os.path.basename(self.library)
+        link_command = self.link_command
+        compile_command = self.compile_command
+        compile_pre = " ".join(list(self.compile_pre))
+        compile_post = " ".join(list(self.compile_post))
+        build = self.build_dir
+        link_flags = " ".join(self.link_options)
+        includes = " ".join(self.includes)
+        sources = " \\\n     ".join(self.sources)
+        with open("Makefile", "w") as fd:
+            print("LIB = %s"%library, file=fd)
+            print("CC = %s"%compile_command, file=fd)
+            print("LINK = %s"%link_command, file=fd)
+            print("CFLAGS = %s %s"%(compile_pre,compile_post), file=fd)
+            print("INCLUDES = %s"%includes, file=fd)
+            print("BUILD = %s"%build, file=fd)
+            print("LINKFLAGS = %s"%link_flags, file=fd)
+            print("SRCS = %s"%sources, file=fd)
+            print("""
+all: $(LIB)
+
+DEPDIR = build/deps
+$(DEPDIR): ; @mkdir -p $@
+
+DEPFILES := $(SRCS:%.cpp=$(DEPDIR)/%.d)
+
+deps: $(DEPFILES)
+
+%/:
+	echo $@
+
+$(DEPDIR)/%.d: %.cpp 
+	mkdir -p $(dir $@)
+	$(CC) $(INCLUDES) -MT $(patsubst $(DEPDIR)%,'$$(BUILD)%',$(patsubst %.d,%.o,$@)) -MM $< -o $@
+
+OBJS = $(addprefix $(BUILD)/, $(SRCS:.cpp=.o))
+
+
+$(BUILD)/%.o: %.cpp
+	mkdir -p $(dir $@)
+	$(CC) $(CFLAGS) $(INCLUDES) -c $< -o $@
+
+
+$(LIB): $(OBJS)
+	$(LINK) $(LINKFLAGS) $(OBJS) -ldl -o $@
+
+
+-include $(DEPFILES)
+""", file=fd)
+
+
+# Customization of the build_ext
+class BuildMakefileCommand(build_ext):
+    """
+    Override some behavior in extension building:
+
+    1. handle compiler flags for different compilers via a dictionary.
+    2. try to disable warning -Wstrict-prototypes is valid for C/ObjC but not for C++
+    """
+
+    # extra compile args
+    copt = {'msvc': [],
```

Comment:
should we unify the cflags handling in a base class, so it does not run out of sync once the central build_ext changes?

Reply:
Yes.  That would be a good idea.   I am curious if the Makefile it is producing is working for you now.  I have pattern for VC as well but it is buggy because I don't have a pattern to make the dep section properly.

Reply:
it fails with some AttributeError on my system:

```
$ python setup.py build_makefile
/home/marscher/miniconda3/envs/jpype_nonp/lib/python3.8/distutils/dist.py:274: UserWarning: Unknown distribution option: 'use_scm_version'
  warnings.warn(msg)
running build_makefile
running build_java
Jar cache is missing, using --enable-build-jar to recreate it.
  ant -Dbuild=../build -f native/build.xml
Buildfile: /home/marscher/sources/jpype/native/build.xml

compile:
    [javac] /home/marscher/sources/jpype/native/build.xml:20: warning: 'includeantruntime' was not set, defaulting to build.sysclasspath=last; set to false for repeatable builds
    [javac] /home/marscher/sources/jpype/native/build.xml:27: warning: 'includeantruntime' was not set, defaulting to build.sysclasspath=last; set to false for repeatable builds

jar:

all:

BUILD SUCCESSFUL
Total time: 0 seconds
running build_thunk
Building thunks
Skip build thunks
building '_jpype' extension
Traceback (most recent call last):
  File "setup.py", line 9, in <module>
    setup(
  File "/home/marscher/miniconda3/envs/jpype_nonp/lib/python3.8/site-packages/setuptools/__init__.py", line 144, in setup
    return distutils.core.setup(**attrs)
  File "/home/marscher/miniconda3/envs/jpype_nonp/lib/python3.8/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/home/marscher/miniconda3/envs/jpype_nonp/lib/python3.8/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/home/marscher/miniconda3/envs/jpype_nonp/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/marscher/miniconda3/envs/jpype_nonp/lib/python3.8/site-packages/setuptools/command/build_ext.py", line 87, in run
    _build_ext.run(self)
  File "/home/marscher/miniconda3/envs/jpype_nonp/lib/python3.8/distutils/command/build_ext.py", line 340, in run
    self.build_extensions()
  File "/home/marscher/sources/jpype/setupext/build_makefile.py", line 195, in build_extensions
    self.compiler.write()
  File "/home/marscher/sources/jpype/setupext/build_makefile.py", line 69, in write
    library = os.path.basename(self.library)
AttributeError: 'Makefile' object has no attribute 'library'
```

Reply:
Okay I will have to find out which disttools is causing the problem.  I checked on Ubuntu under windows and it works so will have to make a wider scan of versions to see if I can find a stable pattern.   

I am not very happy with Python disttools.   Their API requires overriding stuff and it doesn't seem very stable.  I can make it do stuff that I need but it almost always needs to go into the machinery  which seems like poor planning on their part.